application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/DefaultLicenseUpdater.java

```diff
@@ -0,0 +1,249 @@
+import javax.inject.Named;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.apache.hc.client5.http.classic.methods.HttpPost;
+import org.apache.hc.client5.http.entity.UrlEncodedFormEntity;
+import org.apache.hc.core5.http.NameValuePair;
+import org.apache.hc.core5.http.message.BasicNameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.slf4j.Logger;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.crypto.BinaryStringEncoder;
+import org.xwiki.extension.ExtensionId;
+import org.xwiki.extension.repository.InstalledExtensionRepository;
+import org.xwiki.instance.InstanceIdManager;
+import org.xwiki.properties.converter.Converter;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.xpn.xwiki.XWikiContext;
+import com.xwiki.licensing.License;
+import com.xwiki.licensing.LicenseManager;
+import com.xwiki.licensing.LicenseUpdater;
+import com.xwiki.licensing.LicensedExtensionManager;
+import com.xwiki.licensing.LicensingConfiguration;
+import com.xwiki.licensing.Licensor;
+import com.xwiki.licensing.internal.helpers.HttpClientUtils;
+
+/**
+ * Default implementation of {@link LicenseUpdater}.
+ *
+ * @version $Id$
+ * @since 1.27
+ */
+@Component
+@Singleton
+public class DefaultLicenseUpdater implements LicenseUpdater
+{
+    private static final String OUTPUT_SYNTAX = "outputSyntax";
+
+    private static final String PLAIN = "plain";
+
+    private static final String DATA = "data";
+
+    private static final String ERROR =
+        "Failed to update license for [{}]. Please contact sales@xwiki.com for eventual problems.";
+
+    private static final String FEATURE_ID = "featureId";
+
+    private static final String INSTANCE_ID = "instanceId";
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private LicensedExtensionManager licensedExtensionManager;
+
+    @Inject
+    private InstalledExtensionRepository installedExtensionRepository;
+
+    @Inject
+    private LicensingConfiguration licensingConfig;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Provider<InstanceIdManager> instanceIdManagerProvider;
+
+    @Inject
+    private Provider<Licensor> licensorProvider;
+
+    @Inject
+    private Provider<LicenseManager> licenseManagerProvider;
+
+    @Inject
+    @Named("Base64")
+    private BinaryStringEncoder base64decoder;
+
+    @Inject
+    private Converter<License> converter;
+
+    @Inject
+    private HttpClientUtils httpUtils;
+
+    @Override
+    public void renewLicense(ExtensionId extensionId)
+    {
+        try {
+            logger.debug("Try renewing the license of [{}], in order to include new found changes.", extensionId);
+
+            String errorMsg = String.format(
+                "Failed to update license for [%s]. Please contact sales@xwiki.com for eventual problems.",
```

Comment:
This looks very similar to the value of the ``ERROR`` static field. Moreover, so far we tried to keep the Licensing Application generic, and put XWiki SAS stuff in the Store Application. Maybe we can keep the message generic or use a translation key whose value can be overwritten from the Store Application.